src/util/subscription-set/expiring-sorted-set.ts

```diff
@@ -16,28 +19,83 @@ interface ExpiringSortedSetEntry<T> {
  * @typeParam T - the type of the set entries' values
  */
 export class ExpiringSortedSet<T> implements SubscriptionSet<T> {
-  map = new Map<string, ExpiringSortedSetEntry<T>>()
+  capacity: number
+  map: Map<string, LinkedListNode<ExpiringSortedSetEntry<T>>>
+  list: DoubleLinkedList
+
+  constructor(capacity: number) {
+    this.capacity = capacity
+    this.map = new Map<string, LinkedListNode<ExpiringSortedSetEntry<T>>>()
+    this.list = new DoubleLinkedList()
+  }
 
   add(key: string, value: T, ttl: number) {
-    this.map.set(key, {
-      value,
-      expirationTimestamp: Date.now() + ttl,
-    })
+    if (this.map.has(key)) {
+      const node = this.map.get(key) as LinkedListNode<ExpiringSortedSetEntry<T>>
```

Comment:
I think to avoid having to cast the object we could rearrange this check to something like this in all of these methods?
```suggestion
    const node = this.map.get(key)
    if (node) {
```